ecs-init/config/common.go

```diff
@@ -140,14 +139,16 @@ var validDrivers = map[string]struct{}{
 
 // GetAgentPartitionBucketRegion returns the s3 bucket region where ECS Agent artifact is located
 func GetAgentPartitionBucketRegion(region string) (string, error) {
-	regionPartition, ok := endpoints.PartitionForRegion(endpoints.DefaultPartitions(), region)
-	if !ok {
-		return "", errors.Errorf("could not resolve partition ID for region %q", region)
+	// If the AWS SDK cannot match the region to a partition, it returns the default partition (aws).
+	// It only returns nil if the default partition is undefined.
+	partition := awsrulesfn.GetPartition(region)
+	if partition == nil {
+		return "", errors.Errorf("could not resolve partition for region %q", region)
 	}
 
-	bucketRegion, ok := partitionBucketRegion[regionPartition.ID()]
+	bucketRegion, ok := partitionBucketRegion[partition.Name]
```

Comment:
Have we confirmed that partition name is guaranteed to always be the same as partition ID? Or has the testing we have done for this pull request include the case where ecs-init downloads Agent from the partitional S3 bucket that stores the Agent in non-aws partition region?

If testing was done only in aws partition, I am concerned we may not have accounted for some potential discrepancy between partition name and ID which would cause L149 here to consistently result in `!ok`. The test case in aws partition would pass even if this discrepancy exists because as you mentioned, ecs-init defaults the region to us-east-1 when call to `GetAgentPartitionBucketRegion` function fails.

Reply:
> Have we confirmed that partition name is guaranteed to always be the same as partition ID?

[This](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partitions.json) is the source file used to generate [partition metadata](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partitions.go). For all existing partitions, ID and name are the same, but that might not be true for future partitions. If this is a concern, I can implement an alternative to [`getPartition`](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partition.go#L33-L57) that returns [`Partition`](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partition.go#L6-L11), which has ID accessible, instead of [`PartitionConfig`](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partition.go#L14-L21). It could also return `nil` instead of the default aws partition when there's no matching partition for the region.

> Or has the testing we have done for this pull request include the case where ecs-init downloads Agent from the partitional S3 bucket that stores the Agent in non-aws partition region?

I've tested `ecs-init` from this PR in aws (us-west-2) and aws-cn (cn-north-1).

Reply:
Somewhat unrelated...we always fallback to `us-east-1` as the download region, but does S3 support cross-partition downloads?

Reply:
> [This](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partitions.json) is the source file used to generate [partition metadata](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partitions.go). For all existing partitions, ID and name are the same, but that might not be true for future partitions. If this is a concern, I can implement an alternative to [`getPartition`](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partition.go#L33-L57) that returns [`Partition`](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partition.go#L6-L11), which has ID accessible, instead of [`PartitionConfig`](https://github.com/aws/aws-sdk-go-v2/blob/main/internal/endpoints/awsrulesfn/partition.go#L14-L21). It could also return `nil` instead of the default aws partition when there's no matching partition for the region.

Thanks for confirming! Yeah I am aligned with doing both of these to persist existing behavior and have clearer visibility into when there is no matching partition.

Reply:
> Somewhat unrelated...we always fallback to `us-east-1` as the download region, but does S3 support cross-partition downloads?

As far as I know S3 does not support cross-partition downloads. So in instances where we fallback to `us-east-1` in a non-aws partition, my understanding is that download would fail in this case.

Reply:
Changes are implemented in https://github.com/aws/amazon-ecs-agent/pull/4372/commits/3f3b0fec1dcfa8750719a755c5f4b2a75268832a.

Reply:
> Somewhat unrelated...we always fallback to us-east-1 as the download region, but does S3 support cross-partition downloads?

Current understanding: S3 cross-partition downloads are possible if the file being downloaded has public read access granted to it (reference steps to do so: https://repost.aws/knowledge-center/read-access-objects-s3-bucket).